File: components/espectre/pca_detector.cpp
/*
 * ESPectre - PCA Detector Implementation
 * 
 * Alternative motion detection algorithm based on PCA and Pearson correlation.
 * 
 * Author: Francesco Pace <francesco.pace@gmail.com>
 * License: GPLv3
 */

#include "pca_detector.h"
#include <cmath>
#include <cstring>
#include <algorithm>
#include "esphome/core/log.h"

namespace esphome {
namespace espectre {

static const char *TAG = "PCADetector";

// ============================================================================
// Utility Functions
// ============================================================================

float pearson_correlation(const float* a, const float* b, size_t len) {
    if (len == 0) return 0.0f;
    
    // Calculate means
    float mean_a = 0.0f, mean_b = 0.0f;
    for (size_t i = 0; i < len; i++) {
        mean_a += a[i];
        mean_b += b[i];
    }
    mean_a /= static_cast<float>(len);
    mean_b /= static_cast<float>(len);
    
    // Calculate covariance and variances
    float cov_sum = 0.0f;
    float var_sum_a = 0.0f;
    float var_sum_b = 0.0f;
    
    for (size_t i = 0; i < len; i++) {
        float diff_a = a[i] - mean_a;
        float diff_b = b[i] - mean_b;
        cov_sum += diff_a * diff_b;
        var_sum_a += diff_a * diff_a;
        var_sum_b += diff_b * diff_b;
    }
    
    float denominator = std::sqrt(var_sum_a * var_sum_b);
    if (denominator < 1e-10f) return 0.0f;
    
    return cov_sum / denominator;
}

bool pca_power_method(float** data_matrix, size_t num_rows, size_t num_cols,
                      float* output, size_t max_iters, float precision) {
    if (num_rows == 0 || num_cols == 0 || !data_matrix || !output) {
        return false;
    }
    
    // Compute covariance matrix (num_rows x num_rows)
    float zoom_out = static_cast<float>(num_rows * num_cols);
    
    // Allocate covariance matrix on stack (small size)
    float cov_matrix[PCA_WINDOW_SIZE][PCA_WINDOW_SIZE] = {0};
    
    for (size_t i = 0; i < num_rows; i++) {
        for (size_t j = 0; j <= i; j++) {
            float cov_sum = 0.0f;
            for (size_t k = 0; k < num_cols; k++) {
                cov_sum += data_matrix[i][k] * data_matrix[j][k];
            }
            cov_matrix[i][j] = cov_sum / zoom_out;
            if (i != j) {
                cov_matrix[j][i] = cov_matrix[i][j];
            }
        }
    }
    
    // Power method to find principal eigenvector
    float eigenvector[PCA_WINDOW_SIZE];
    for (size_t i = 0; i < num_rows; i++) {
        eigenvector[i] = 1.0f;
    }
    
    float eigenvalue = 1.0f;
    float eigenvalue_last = 0.0f;
    
    for (size_t iter = 0; iter < max_iters; iter++) {
        if (std::fabs(eigenvalue - eigenvalue_last) <= precision) {
            break;
        }
        
        eigenvalue_last = eigenvalue;
        eigenvalue = 0.0f;
        
        // Multiply: new_vec = cov_matrix @ eigenvector
        float new_vec[PCA_WINDOW_SIZE] = {0};
        for (size_t i = 0; i < num_rows; i++) {
            for (size_t j = 0; j < num_rows; j++) {
                new_vec[i] += cov_matrix[i][j] * eigenvector[j];
            }
            if (new_vec[i] > eigenvalue) {
                eigenvalue = new_vec[i];
            }
        }
        
        // Normalize
        if (eigenvalue > 1e-10f) {
            for (size_t i = 0; i < num_rows; i++) {
                eigenvector[i] = new_vec[i] / eigenvalue;
            }
        }
    }
    
    // Project data onto eigenvector to get principal component
    for (size_t k = 0; k < num_cols; k++) {
        output[k] = 0.0f;
        for (size_t i = 0; i < num_rows; i++) {
            output[k] += data_matrix[i][k] * eigenvector[i];
        }
        output[k] /= static_cast<float>(num_rows);
    }
    
    return true;
}

// ============================================================================
// PCADetector Implementation
// ============================================================================

PCADetector::PCADetector()
    : state_(MotionState::IDLE)
    , csi_buffer_count_(0)
    , pca_buffer_count_(0)
    , pca_buffer_idx_(0)
    , calibration_count_(0)
    , jitter_buffer_count_(0)
    , jitter_buffer_idx_(0)
    , current_jitter_(0.0f)
    , current_wander_(0.0f)
    , threshold_(PCA_DEFAULT_THRESHOLD)
    , packet_count_(0)
    , last_detection_result_(false)
    , threshold_externally_set_(false) {
    
    // Allocate buffers
    for (size_t i = 0; i < PCA_WINDOW_SIZE; i++) {
        csi_buffer_[i] = new float[PCA_NUM_SUBCARRIERS]();
    }
    for (size_t i = 0; i < PCA_BUFFER_SIZE; i++) {
        pca_buffer_[i] = new float[PCA_NUM_SUBCARRIERS]();
    }
    for (size_t i = 0; i < PCA_CALIBRATION_SAMPLES; i++) {
        calibration_data_[i] = new float[PCA_NUM_SUBCARRIERS]();
    }
    
    std::memset(jitter_buffer_, 0, sizeof(jitter_buffer_));
    
    ESP_LOGI(TAG, "Initialized (subcarriers=%zu, threshold=%.3f)", PCA_NUM_SUBCARRIERS, threshold_);
}

PCADetector::~PCADetector() {
    for (size_t i = 0; i < PCA_WINDOW_SIZE; i++) {
        delete[] csi_buffer_[i];
    }
    for (size_t i = 0; i < PCA_BUFFER_SIZE; i++) {
        delete[] pca_buffer_[i];
    }
    for (size_t i = 0; i < PCA_CALIBRATION_SAMPLES; i++) {
        delete[] calibration_data_[i];
    }
}

// ============================================================================
// IDetector Interface
// ============================================================================

void PCADetector::process_packet(const int8_t* csi_data, size_t csi_len,
                                  const uint8_t* /* selected_subcarriers */,
                                  uint8_t /* num_subcarriers */) {
    // PCA uses its own subcarrier selection (every PCA_SUBCARRIER_STEP-th)
    // Ignores the passed subcarrier selection
    
    packet_count_++;
    
    // Extract amplitudes (every PCA_SUBCARRIER_STEP-th subcarrier)
    size_t buffer_idx = csi_buffer_count_ % PCA_WINDOW_SIZE;
    extract_amplitudes(csi_data, csi_len, csi_buffer_[buffer_idx]);
    csi_buffer_count_++;
    
    // Need enough data for PCA
    if (csi_buffer_count_ < PCA_WINDOW_SIZE) {
        current_jitter_ = 0.0f;
        current_wander_ = 0.0f;
        last_detection_result_ = false;
        return;
    }
    
    // Compute PCA on current window
    float pca_current[PCA_NUM_SUBCARRIERS];
    
    // Build pointer array for pca_power_method
    float* window_ptrs[PCA_WINDOW_SIZE];
    for (size_t i = 0; i < PCA_WINDOW_SIZE; i++) {
        size_t idx = (csi_buffer_count_ - PCA_WINDOW_SIZE + i) % PCA_WINDOW_SIZE;
        window_ptrs[i] = csi_buffer_[idx];
    }
    
    if (!pca_power_method(window_ptrs, PCA_WINDOW_SIZE, PCA_NUM_SUBCARRIERS, pca_current)) {
        current_jitter_ = 0.0f;
        current_wander_ = 0.0f;
        last_detection_result_ = false;
        return;
    }
    
    // Calculate waveform metrics (as correlations, 0-1 range)
    float jitter_corr = compute_jitter(pca_current);
    float wander_corr = compute_wander(pca_current);
    
    // Invert: 1 - correlation, so high value = movement/change
    // Scale by PCA_SCALE (1000) to match MVS threshold range (0.1-10.0)
    float jitter_inverted = (1.0f - jitter_corr) * PCA_SCALE;
    float wander_inverted = (1.0f - wander_corr) * PCA_SCALE;
    
    current_jitter_ = jitter_inverted;
    current_wander_ = wander_inverted;
    
    // Store PCA vector for future comparisons
    std::memcpy(pca_buffer_[pca_buffer_idx_], pca_current, PCA_NUM_SUBCARRIERS * sizeof(float));
    pca_buffer_idx_ = (pca_buffer_idx_ + 1) % PCA_BUFFER_SIZE;
    pca_buffer_count_++;
    
    // Collect calibration samples during initial phase (for wander calculation)
    if (calibration_count_ < PCA_CALIBRATION_SAMPLES && pca_buffer_count_ % 5 == 0) {
        std::memcpy(calibration_data_[calibration_count_], pca_current, PCA_NUM_SUBCARRIERS * sizeof(float));
        calibration_count_++;
    }
    
    // Add to jitter buffer for smoothing
    jitter_buffer_[jitter_buffer_idx_] = jitter_inverted;
    jitter_buffer_idx_ = (jitter_buffer_idx_ + 1) % PCA_JITTER_BUFFER_SIZE;
    if (jitter_buffer_count_ < PCA_JITTER_BUFFER_SIZE) {
        jitter_buffer_count_++;
    }
    
    // Need enough jitter samples for detection
    if (jitter_buffer_count_ < PCA_MOVE_BUFFER_SIZE) {
        last_detection_result_ = false;
        return;
    }
    
    // Count threshold violations in recent window
    size_t move_count = 0;
    
    // Calculate median of jitter buffer
    float sorted_jitter[PCA_JITTER_BUFFER_SIZE];
    std::memcpy(sorted_jitter, jitter_buffer_, jitter_buffer_count_ * sizeof(float));
    std::sort(sorted_jitter, sorted_jitter + jitter_buffer_count_);
    float jitter_median = sorted_jitter[jitter_buffer_count_ / 2];
    
    for (size_t i = 0; i < PCA_MOVE_BUFFER_SIZE; i++) {
        size_t idx = (jitter_buffer_idx_ + PCA_JITTER_BUFFER_SIZE - 1 - i) % PCA_JITTER_BUFFER_SIZE;
        float jitter_val = jitter_buffer_[idx];
        
        // Dual condition (values are scaled by PCA_SCALE)
        if (jitter_val > threshold_ || 
            (jitter_val > jitter_median && jitter_val > 10.0f)) {
            move_count++;
        }
    }
    
    // Motion detected if enough violations
    last_detection_result_ = (move_count >= PCA_OUTLIERS_NUM);
}

void PCADetector::update_state() {
    // Update state based on last detection result
    if (last_detection_result_) {
        state_ = MotionState::MOTION;
    } else {
        state_ = MotionState::IDLE;
    }
}

bool PCADetector::set_threshold(float threshold) {
    // Threshold is now scaled by PCA_SCALE (1000), so valid range is 0.0-10.0
    // matching the MVS threshold range (SEGMENTATION_MIN/MAX_THRESHOLD)
    if (threshold < 0.0f || threshold > 10.0f) {
        ESP_LOGE(TAG, "Invalid threshold: %.3f (must be 0.0-10.0)", threshold);
        return false;
    }
    threshold_ = threshold;
    threshold_externally_set_ = true;
    ESP_LOGI(TAG, "Threshold updated: %.4f", threshold);
    return true;
}

void PCADetector::reset() {
    state_ = MotionState::IDLE;
    csi_buffer_count_ = 0;
    pca_buffer_count_ = 0;
    pca_buffer_idx_ = 0;
    calibration_count_ = 0;
    jitter_buffer_count_ = 0;
    jitter_buffer_idx_ = 0;
    current_jitter_ = 0.0f;
    current_wander_ = 0.0f;
    packet_count_ = 0;
    last_detection_result_ = false;
    // Note: threshold_ and threshold_externally_set_ are preserved across reset
    
    ESP_LOGD(TAG, "Reset");
}

// ============================================================================
// Private Methods
// ============================================================================

void PCADetector::extract_amplitudes(const int8_t* csi_data, size_t len, float* amplitudes) {
    size_t max_sc = len / 2;  // I/Q pairs
    size_t out_idx = 0;
    
    for (size_t sc = 0; sc < max_sc && out_idx < PCA_NUM_SUBCARRIERS; sc += PCA_SUBCARRIER_STEP) {
        // CSI format: [Q, I, Q, I, ...] (Imaginary first, then Real)
        float q = static_cast<float>(csi_data[sc * 2]);
        float i = static_cast<float>(csi_data[sc * 2 + 1]);
        amplitudes[out_idx++] = std::sqrt(i * i + q * q);
    }
    
    // Fill remaining with zeros if needed
    while (out_idx < PCA_NUM_SUBCARRIERS) {
        amplitudes[out_idx++] = 0.0f;
    }
}

float PCADetector::compute_jitter(const float* pca_current) {
    if (pca_buffer_count_ < 2) {
        return 1.0f;  // No past data, assume max correlation (no movement)
    }
    
    float max_corr = 0.0f;
    size_t num_to_check = std::min(pca_buffer_count_ - 1, PCA_MOVE_BUFFER_SIZE - 1);
    
    for (size_t i = 0; i < num_to_check; i++) {
        size_t past_idx = (pca_buffer_idx_ + PCA_BUFFER_SIZE - 1 - i) % PCA_BUFFER_SIZE;
        float corr = std::fabs(pearson_correlation(pca_current, pca_buffer_[past_idx], PCA_NUM_SUBCARRIERS));
        if (corr > max_corr) {
            max_corr = corr;
        }
    }
    
    return max_corr;
}

float PCADetector::compute_wander(const float* pca_current) {
    if (calibration_count_ == 0) {
        return 1.0f;
    }
    
    float max_corr = 0.0f;
    for (size_t i = 0; i < calibration_count_; i++) {
        float corr = std::fabs(pearson_correlation(pca_current, calibration_data_[i], PCA_NUM_SUBCARRIERS));
        if (corr > max_corr) {
            max_corr = corr;
        }
    }
    
    return max_corr;
}

}  // namespace espectre
}  // namespace esphome

File: micro-espectre/src/pca_detector.py
"""
Micro-ESPectre - PCA Detector

Principal Component Analysis detector implementation.
Based on Espressif's esp_radar algorithm.

Author: Francesco Pace <francesco.pace@gmail.com>
License: GPLv3
"""
import math

try:
    from src.detector_interface import IDetector, MotionState
except ImportError:
    from detector_interface import IDetector, MotionState


def pearson_correlation(a, b):
    """
    Calculate Pearson correlation coefficient between two vectors.
    
    Returns value in range [-1, 1], where:
    - 1 = perfectly correlated (identical signal shape)
    - 0 = no correlation
    - -1 = perfectly anti-correlated
    """
    if len(a) != len(b) or len(a) == 0:
        return 0.0
    
    n = len(a)
    mean_a = sum(a) / n
    mean_b = sum(b) / n
    
    cov_sum = 0.0
    var_sum_a = 0.0
    var_sum_b = 0.0
    
    for i in range(n):
        diff_a = a[i] - mean_a
        diff_b = b[i] - mean_b
        cov_sum += diff_a * diff_b
        var_sum_a += diff_a ** 2
        var_sum_b += diff_b ** 2
    
    denominator = math.sqrt(var_sum_a * var_sum_b)
    if denominator < 1e-10:
        return 0.0
    
    return cov_sum / denominator


def pca_power_method(data_matrix, max_iters=30, precision=0.0001):
    """
    PCA using power method to find principal eigenvector.
    Replicates Espressif's pca() function.
    
    Args:
        data_matrix: 2D list of shape (num_packets, num_subcarriers)
        max_iters: Maximum iterations for power method
        precision: Convergence threshold
    
    Returns:
        Principal component vector or None
    """
    if len(data_matrix) == 0:
        return None
    
    rows = len(data_matrix[0])  # num_subcarriers
    cols = len(data_matrix)      # num_packets
    
    if cols == 0 or rows == 0:
        return None
    
    # Transpose: subcarriers as rows, packets as columns
    matrix = [[data_matrix[j][i] for j in range(cols)] for i in range(rows)]
    
    # Compute covariance matrix (cols x cols)
    zoom_out = rows * cols
    cov_matrix = [[0.0] * cols for _ in range(cols)]
    
    for i in range(cols):
        for j in range(i + 1):
            cov_sum = 0.0
            for k in range(rows):
                cov_sum += matrix[k][i] * matrix[k][j]
            cov_matrix[i][j] = cov_sum / zoom_out
            if i != j:
                cov_matrix[j][i] = cov_matrix[i][j]
    
    # Power method to find principal eigenvector
    eigenvector = [1.0] * cols
    eigenvalue = 1.0
    eigenvalue_last = 0.0
    
    for _ in range(max_iters):
        if abs(eigenvalue - eigenvalue_last) <= precision:
            break
        
        eigenvalue_last = eigenvalue
        eigenvalue = 0.0
        
        # Multiply: eigenvalue_list = cov_matrix @ eigenvector
        eigenvalue_list = [0.0] * cols
        for i in range(cols):
            for j in range(cols):
                eigenvalue_list[i] += cov_matrix[i][j] * eigenvector[j]
            if eigenvalue_list[i] > eigenvalue:
                eigenvalue = eigenvalue_list[i]
        
        # Normalize
        if eigenvalue > 1e-10:
            eigenvector = [v / eigenvalue for v in eigenvalue_list]
    
    # Project data onto eigenvector
    output = [0.0] * rows
    for i in range(rows):
        for j in range(cols):
            output[i] += matrix[i][j] * eigenvector[j]
        output[i] /= cols
    
    return output


class PCADetector(IDetector):
    """
    PCA-based motion detector (Espressif esp_radar style).
    
    Algorithm:
    1. Collect CSI amplitudes into sliding window
    2. Apply PCA (power method) to extract principal component
    3. Calculate correlation with past PCA vectors (jitter)
    4. Calculate correlation with calibration vectors (wander)
    5. Invert values: output = 1 - correlation (high = movement)
    6. Apply count-based detection for final decision
    
    Note: Threshold must be set externally via set_threshold() after
    calibration with PCACalibrator. Formula: threshold = 1 - min(correlation)
    
    Based on esp-radar component v0.3.1 (Apache-2.0 license).
    """
    
    # Parameters (aligned with C++ implementation)
    DEFAULT_PCA_WINDOW_SIZE = 10
    DEFAULT_MOVE_BUFFER_SIZE = 5
    DEFAULT_OUTLIERS_NUM = 2
    DEFAULT_SUBCARRIER_STEP = 4
    BUFF_MAX_LEN = 25
    PCA_CALIBRATION_SAMPLES = 10  # Same as C++ PCA_CALIBRATION_SAMPLES
    
    # Scale factor for PCA values to match MVS threshold range (0.1-10.0)
    # PCA jitter is ~0.0001-0.001, scaling by 1000 brings it to ~0.1-1.0
    PCA_SCALE = 1000.0
    DEFAULT_MOVE_THRESHOLD = 10.0  # 0.01 * PCA_SCALE
    
    def __init__(self,
                 pca_window_size=DEFAULT_PCA_WINDOW_SIZE,
                 move_buffer_size=DEFAULT_MOVE_BUFFER_SIZE,
                 outliers_num=DEFAULT_OUTLIERS_NUM,
                 threshold=DEFAULT_MOVE_THRESHOLD,
                 subcarrier_step=DEFAULT_SUBCARRIER_STEP):
        """
        Initialize PCA detector.
        
        Args:
            pca_window_size: Packets for PCA computation (default: 10)
            move_buffer_size: Buffer for jitter values (default: 5)
            outliers_num: Violations needed to trigger (default: 2)
            threshold: Threshold on inverted jitter (default: 10.0, scaled by PCA_SCALE)
            subcarrier_step: Use every Nth subcarrier (default: 4)
        """
        self.pca_window_size = pca_window_size
        self.move_buffer_size = move_buffer_size
        self.outliers_num = outliers_num
        self._threshold = threshold
        self.subcarrier_step = subcarrier_step
        
        # Buffers
        self.csi_buffer = []
        self.pca_buffer = []
        self.pca_count = 0
        self.calibration_data = []  # For wander calculation (max PCA_CALIBRATION_SAMPLES)
        self.jitter_buffer = []
        
        # State
        self._state = MotionState.IDLE
        self._packet_count = 0
        self._motion_count = 0
        self._current_jitter = 0.0
        self._current_wander = 0.0
        self._threshold_externally_set = False  # Same as C++ threshold_externally_set_
        
        # Tracking
        self.jitter_history = []
        self.wander_history = []
        self.state_history = []
        self.track_data = False
    
    def _extract_amplitudes(self, csi_data):
        """Extract amplitudes using step-based subcarrier selection."""
        amplitudes = []
        max_sc = min(64, len(csi_data) // 2)
        
        for sc_idx in range(0, max_sc, self.subcarrier_step):
            i = sc_idx * 2
            if i + 1 < len(csi_data):
                # Espressif CSI format: [Imaginary, Real]
                imag = float(csi_data[i])
                real = float(csi_data[i + 1])
                amplitudes.append(math.sqrt(real * real + imag * imag))
        
        return amplitudes
    
    def _compute_waveform_jitter(self, pca_current):
        """
        Compute jitter as max correlation with RECENT PCA vectors.
        
        Jitter measures short-term changes in the signal.
        High correlation (close to 1) = stable = low jitter
        Low correlation = changing = high jitter
        
        Note: This compares with pca_buffer (recent), NOT calibration_data.
        Wander uses calibration_data for baseline comparison.
        """
        # Need at least 2 past vectors (aligned with C++ pca_buffer_count_ < 2)
        if len(self.pca_buffer) < 2:
            return 1.0  # No past data, assume max correlation (no movement)
        
        max_corr = 0.0
        # Compare with most recent past PCA vectors
        # Use MOVE_BUFFER_SIZE - 1 to match C++ behavior
        num_compare = min(self.move_buffer_size - 1, len(self.pca_buffer) - 1)
        for i in range(num_compare):
            past_idx = len(self.pca_buffer) - 1 - i  # -1 because pca_current not yet appended
            if past_idx < 0:
                break
            pca_past = self.pca_buffer[past_idx]
            corr_val = abs(pearson_correlation(pca_current, pca_past))
            if corr_val > max_corr:
                max_corr = corr_val
        
        return max_corr
    
    def _compute_waveform_wander(self, pca_current):
        """Compute wander as max correlation with calibration samples."""
        if len(self.calibration_data) == 0:
            return 1.0
        
        max_corr = 0.0
        for calib_pca in self.calibration_data:
            corr_val = abs(pearson_correlation(pca_current, calib_pca))
            if corr_val > max_corr:
                max_corr = corr_val
        
        return max_corr
    
    def process_packet(self, csi_data, selected_subcarriers=None):
        """
        Process a CSI packet.
        
        Note: PCA ignores selected_subcarriers, uses its own step-based selection.
        """
        self._packet_count += 1
        
        # Extract amplitudes
        amplitudes = self._extract_amplitudes(csi_data)
        
        # Add to CSI buffer
        self.csi_buffer.append(amplitudes)
        if len(self.csi_buffer) > self.pca_window_size:
            self.csi_buffer.pop(0)
        
        # Need enough data for PCA
        if len(self.csi_buffer) < self.pca_window_size:
            self._current_jitter = 0.0
            self._current_wander = 0.0
            return
        
        # Compute PCA
        pca_current = pca_power_method(self.csi_buffer)
        if pca_current is None:
            self._current_jitter = 0.0
            self._current_wander = 0.0
            return
        
        # Calculate waveform metrics
        jitter_corr = self._compute_waveform_jitter(pca_current)
        wander_corr = self._compute_waveform_wander(pca_current)
        
        # Invert: 1 - correlation, scaled by PCA_SCALE to match MVS range
        self._current_jitter = (1.0 - jitter_corr) * self.PCA_SCALE
        self._current_wander = (1.0 - wander_corr) * self.PCA_SCALE
        
        # Store PCA vector
        self.pca_buffer.append(pca_current[:])
        if len(self.pca_buffer) > self.BUFF_MAX_LEN:
            self.pca_buffer.pop(0)
        self.pca_count += 1
        
        # Collect calibration samples for wander (same as C++ PCA_CALIBRATION_SAMPLES)
        if len(self.calibration_data) < self.PCA_CALIBRATION_SAMPLES and self.pca_count % 5 == 0:
            self.calibration_data.append(pca_current[:])
        
        # Add to jitter buffer
        self.jitter_buffer.append(self._current_jitter)
        if len(self.jitter_buffer) > self.BUFF_MAX_LEN:
            self.jitter_buffer.pop(0)
        
        if self.track_data:
            self.jitter_history.append(self._current_jitter)
            self.wander_history.append(self._current_wander)
    
    def update_state(self):
        """Update motion state based on jitter buffer."""
        if len(self.jitter_buffer) < self.move_buffer_size:
            self._state = MotionState.IDLE
            if self.track_data:
                self.state_history.append('IDLE')
            return self._get_metrics()
        
        # Count threshold violations
        move_count = 0
        
        # Calculate median of ENTIRE jitter buffer (matches C++ behavior)
        sorted_buffer = sorted(self.jitter_buffer)
        jitter_median = sorted_buffer[len(sorted_buffer) // 2]
        
        # Check only the last move_buffer_size values
        jitter_values = self.jitter_buffer[-self.move_buffer_size:]
        for jitter_val in jitter_values:
            # Dual condition (values are scaled by PCA_SCALE)
            if (jitter_val > self._threshold or
                (jitter_val > jitter_median and jitter_val > 10.0)):
                move_count += 1
        
        # Update state
        if move_count >= self.outliers_num:
            self._state = MotionState.MOTION
            self._motion_count += 1
        else:
            self._state = MotionState.IDLE
        
        if self.track_data:
            state_str = 'MOTION' if self._state == MotionState.MOTION else 'IDLE'
            self.state_history.append(state_str)
        
        return self._get_metrics()
    
    def _get_metrics(self):
        """Get current metrics dict."""
        return {
            'jitter': self._current_jitter,
            'wander': self._current_wander,
            'threshold': self._threshold,
            'state': self._state
        }
    
    def get_state(self):
        """Get current motion state."""
        return self._state
    
    def get_motion_metric(self):
        """Get current jitter value."""
        return self._current_jitter
    
    def get_threshold(self):
        """Get current threshold."""
        return self._threshold
    
    def set_threshold(self, threshold):
        """Set detection threshold (must be called after PCACalibrator completes).
        
        Threshold is now scaled by PCA_SCALE (1000), valid range is 0.0-10.0
        matching the MVS threshold range.
        """
        if 0.0 <= threshold <= 10.0:
            self._threshold = threshold
            self._threshold_externally_set = True
            return True
        return False
    
    def is_ready(self):
        """Check if threshold has been set externally (via PCACalibrator)."""
        return self._threshold_externally_set
    
    def reset(self):
        """Reset detector state (threshold is preserved)."""
        self.csi_buffer = []
        self.pca_buffer = []
        self.pca_count = 0
        self.calibration_data = []
        self.jitter_buffer = []
        self._state = MotionState.IDLE
        self._motion_count = 0
        self._current_jitter = 0.0
        self._current_wander = 0.0
        # Note: _threshold and _threshold_externally_set are preserved across reset
        self.jitter_history = []
        self.wander_history = []
        self.state_history = []
    
    def get_name(self):
        """Get detector name."""
        return "PCA"
    
    @property
    def total_packets(self):
        """Total packets processed."""
        return self._packet_count
    
    def get_motion_count(self):
        """Get number of motion detections."""
        return self._motion_count
